plugin/storage/kafka/writer.go

```diff
@@ -0,0 +1,84 @@
+// Copyright (c) 2018 The Jaeger Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package kafka
+
+import (
+	"github.com/Shopify/sarama"
+	"github.com/uber/jaeger-lib/metrics"
+
+	"github.com/jaegertracing/jaeger/model"
+)
+
+type spanWriterMetrics struct {
+	SpansWrittenSuccess metrics.Counter
+	SpansWrittenFailure metrics.Counter
+}
+
+// SpanWriter writes spans to kafka. Implements spanstore.Writer
+type SpanWriter struct {
+	metrics    spanWriterMetrics
+	producer   sarama.AsyncProducer
+	marshaller Marshaller
+	topic      string
+}
+
+// NewSpanWriter initiates and returns a new kafka spanwriter
+func NewSpanWriter(producer sarama.AsyncProducer, topic string, factory metrics.Factory) *SpanWriter {
+	writeMetrics := spanWriterMetrics{
+		SpansWrittenSuccess: factory.Counter("kafka_spans_written", map[string]string{"status": "success"}),
+		SpansWrittenFailure: factory.Counter("kafka_spans_written", map[string]string{"status": "failure"}),
+	}
+
+	go func() {
+		for range producer.Successes() {
+			writeMetrics.SpansWrittenSuccess.Inc(1)
+		}
+	}()
+	go func() {
+		for range producer.Errors() {
+			writeMetrics.SpansWrittenFailure.Inc(1)
+		}
+	}()
+
+	return &SpanWriter{
+		producer:   producer,
+		marshaller: newThriftMarshaller(),
+		topic:      topic,
+		metrics:    writeMetrics,
+	}
+}
+
+// WriteSpan writes the span to kafka.
+func (w *SpanWriter) WriteSpan(span *model.Span) error {
+	spanBytes, err := w.marshaller.Marshal(span)
+	if err != nil {
+		w.metrics.SpansWrittenFailure.Inc(1)
+		return err
+	}
+
+	// The AsyncProducer accepts messages on a channel and produces them asynchronously
+	// in the background as efficiently as possible
+	w.producer.Input() <- &sarama.ProducerMessage{
+		Topic: w.topic,
+		Key:   sarama.StringEncoder(span.TraceID.String()),
+		Value: sarama.ByteEncoder(spanBytes),
+	}
+	return nil
+}
+
+// Close closes SpanWriter by closing producer
```

Comment:
`// Close shuts down SpanWriter by closing the Kafka producer, which closes its channels causing go-routines to exit.`

There may be a race condition here if producer closes the input channel but someone calls WriteSpan after that. May need to keep an atomic flag to stop writes after Close()

Reply:
If the input channel is closed, shouldn't the send to the closed channel panic? 
I think we should revisit how releasing resources work, and do something similar to the following:
- Set healthcheck to unavailable
- Close tchannel and http servers
- Close span writers

Reply:
I'm gonna allow this to land and punt on this. This rarely is a concern given how we do deploys internally anyway. And who does graceful shutdowns anyway.